src/DefinitionResolver.php

```diff
@@ -479,8 +483,10 @@ public function resolveExpressionNodeToType(Node\Expr $expr): Type
                 if ($expr instanceof Node\Expr\MethodCall) {
                     $fqn .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
+                if ($def = $this->index->getDefinition($fqn)) {
+                    if ($def->type instanceof Types\This || $def->type instanceof Types\Self_) {
+                        return $this->resolveExpressionNodeToType($expr->var);
+                    }
```

Comment:
Because the `$this` reference may not come from a variable. 
If the check is removed the following example no longer works:
```php
<?php

class ThisClassPrefix extends TestClass
{
    private $foo;
    private $bar;

    /**
     * @return self
     */
    protected function method()
    {
        return $this;
    }
    public function test()
    {
        $this->method()->|
    }
}
```